### PR TITLE
BG-17474 Initial Tezos KeyPair code

### PR DESCRIPTION
**Pull request by argjv**
[BG-17474](https://bitgoinc.atlassian.net/browse/BG-17474) Add Tezos key pair management logic.

### CODE REVIEW

File: src/coin/xtz/keyPair.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  getKeys(): DefaultKeys {
+    // Always use the compressed version to be consistent
+    const pub = this.keyPair.Q.getEncoded(true);
+
+    const result: DefaultKeys = {
+      pub: Utils.base58encode(Utils.prefix.sppk, pub),
+    };
+
+    if (this.keyPair.d) {
+      const prv = this.keyPair.getPrivateKeyBuffer();
```
Comment (OttoAllmendinger) [inline src/coin/xtz/keyPair.ts]: this will be deprecated, alternative was added in bitgo-utxo-lib 1.8.0 https://github.com/BitGo/bitgo-utxo-lib/releases/tag/1.8.0

File: src/coin/xtz/keyPair.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+   * @param prv An extended or raw private key
+   */
+  private recordKeysFromPrivateKey(prv: string): void {
+    if (Crypto.isValidXprv(prv)) {
+      this.hdNode = HDNode.fromBase58(prv);
+    } else if (Crypto.isValidPrv(prv)) {
+      // Cannot create the HD node without the chain code, so create a regular Key Chain
+      this.keyPair = ECPair.fromPrivateKeyBuffer(new Buffer(prv, 'hex'));
+    } else if (Utils.isValidTezosKey(Utils.prefix.spsk, prv)) {
+      this.keyPair = ECPair.fromPrivateKeyBuffer(Utils.decodeKey(Utils.prefix.spsk, prv));
```
Comment (OttoAllmendinger) [inline src/coin/xtz/keyPair.ts]: https://github.com/BitGo/bitgo-utxo-lib/releases/tag/1.8.0

File: src/coin/xtz/keyPair.ts (hunk 1 above)
Comment (argjv) [reply src/coin/xtz/keyPair.ts]: Will it be deprecated or is it already? if it's not done yet, no point on delaying the PR on that. If it is deprecated already, do you have a test or method link to use as reference?

Comment (OttoAllmendinger) [reply src/coin/xtz/keyPair.ts]: It will be deprecated in the future. The release notes describe an alternative method that was added in 1.8.0 which is supported going forward.

I'm not NACKing this PR, this is just a tip to save work in the future.

File: src/coin/xtz/keyPair.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    return result;
+  }
+
+  /**
+   * Get a public address in the specified format, or in base58 if none is provided.
+   */
+  getAddress(): string {
+    const { pub } = this.getKeys();
+    const out = Buffer.alloc(20);
+    const b2b = blake2b(out.length).update(pub).digest(out);
```
Comment (CapnMigraine) [inline src/coin/xtz/keyPair.ts]: nit: use eslint's suggested format:
```
    const b2b = blake2b(out.length)
      .update(pub)
      .digest(out);

File: src/coin/xtz/keyPair.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+  /**
+   * Get the extended public key, and the private key if one is available. This is only possible
+   * when the key pair was created from a seed or extended keys.
+   */
+  getExtendedKeys(): ExtendedKeys {
+    if (!this.hdNode) {
+      throw new Error('Cannot get extended keys');
+    }
+    let result: ExtendedKeys = {
```
Comment (CapnMigraine) [inline src/coin/xtz/keyPair.ts]: nit: use `const`

File: src/coin/xtz/keyPair.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
+import { DefaultKeys, ExtendedKeys} from '../baseCoin/iface';
```
Comment (CapnMigraine) [inline src/coin/xtz/keyPair.ts]: nit: add space before `}`

File: src/coin/xtz/iface.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+/**
+ * A Tezos private key with the spsk prefix or raw
+ */
+export type PrivateKey = {
+  prv: string
```
Comment (CapnMigraine) [inline src/coin/xtz/iface.ts]: nit: add missing `;`

File: src/coin/xtz/iface.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+/**
+ * A Tezos public key with the sppk prefix or raw
+ */
+export type PublicKey = {
+  pub: string
```
Comment (CapnMigraine) [inline src/coin/xtz/iface.ts]: nit: same as above

File: src/coin/xtz/keyPair.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  private recordKeysFromPublicKey(pub: string): void {
+    if (Crypto.isValidXpub(pub)) {
+      this.hdNode = HDNode.fromBase58(pub);
+    } else if (Crypto.isValidPub(pub)) {
+      // Cannot create an HD node without the chain code, so create a regular Key Chain
+      this.keyPair = ECPair.fromPublicKeyBuffer(new Buffer(pub, 'hex'));
+    } else if (Utils.isValidTezosKey(Utils.prefix.sppk, pub)) {
+      this.keyPair = ECPair.fromPublicKeyBuffer(Utils.decodeKey(Utils.prefix.sppk, pub));
+    } else {
+      throw new Error('Unsupported public key: ' + pub );
```
Comment (CapnMigraine) [inline src/coin/xtz/keyPair.ts]: nit: remove space before `)`

File: src/coin/xtz/utils.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
+import * as base58check from 'bs58check';
+
+/**
+ * Encode the payload to base58 with a specific Tezos prefix.
+ * @param {Uint8Array} prefix to add to the encoded payload
+ * @param {Buffer} payload to encode
+ * @return {any} base58 payload with a Tezos prefix
+ */
+export function base58encode(prefix: Uint8Array, payload: Buffer) {
```
Comment (CapnMigraine) [inline src/coin/xtz/utils.ts]: nit: add return type

File: src/coin/xtz/utils.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  let n = Buffer.alloc(prefix.length + payload.length);
```
Comment (CapnMigraine) [inline src/coin/xtz/utils.ts]: nit: you can use `const`

File: src/coin/xtz/utils.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  n.set(prefix);
+  n.set(payload, prefix.length);
+  // @ts-ignore
```
Comment (CapnMigraine) [inline src/coin/xtz/utils.ts]: why is this needed here?

File: src/coin/xtz/utils.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  return base58check.encode(n);
+}
+
+/** Whether or not the string is a valid Tezos key given a prefix. */
+export function isValidTezosKey(prefix: Uint8Array, text: string) {
```
Comment (CapnMigraine) [inline src/coin/xtz/utils.ts]: nit: add return type

File: src/index.ts [hunk 13]
```diff
@@ -10,8 +10,13 @@ export { BaseCoin };
...
 export { Trx };
 
-// TODO: verify these with the coins available in the statics lib
-export const supportedCoins = ['trx', 'ttrx'];
+import * as Xtz from './coin/xtz';
+export { Xtz };
+
+export const supportedCoins = {
+  'trx': new Trx.TransactionBuilder(coins.get('trx')),
+  'ttrx': new Trx.TransactionBuilder(coins.get('ttrx'))
```
Comment (CapnMigraine) [inline src/index.ts]: nit: remove single quotes from keys and add dangling comma in line 18

File: src/index.ts [hunk 14]
```diff
@@ -21,9 +26,9 @@ export const supportedCoins = ['trx', 'ttrx'];
  */
 export function getBuilder(coinName: string) {
```
Comment (CapnMigraine) [inline src/index.ts]: nit: add return type

File: src/coin/xtz/iface.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
+/**
+ * A Tezos private key with the spsk prefix or raw
+ */
+export type PrivateKey = {
```
Comment (twtaylorbitgo) [inline src/coin/xtz/iface.ts]: doesn't use a couple of different types of prvs? is spsk a curve or the key type?

File: src/coin/xtz/keyPair.ts [hunk 16]
```diff
@@ -0,0 +1,128 @@
+import { DefaultKeys, ExtendedKeys} from '../baseCoin/iface';
+import * as crypto from 'crypto';
+import { HDNode, ECPair } from 'bitgo-utxo-lib';
+import * as Utils from './utils';
+import { isPrivateKey, isPublicKey, isSeed, KeyPairOptions } from './iface';
+import * as Crypto from '../../utils/crypto';
```
Comment (twtaylorbitgo) [inline src/coin/xtz/keyPair.ts]: maybe name this cryptoUtils instead, its confusing with the crypto import above

File: src/coin/xtz/utils.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+  const decodedText = base58check.decode(text);
+  const textPrefix = decodedText.slice(0, prefix.length);
```
Comment (twtaylorbitgo) [inline src/coin/xtz/utils.ts]: we should have a check for the length here

Comment (twtaylorbitgo) [reply src/coin/xtz/utils.ts]: actually, i was concerned about the 0 case but it seems like it would just return false here around 22

File: src/coin/xtz/keyPair.ts (hunk 1 above)
Comment (OttoAllmendinger) [reply src/coin/xtz/keyPair.ts]: the util method is forward-compatible
